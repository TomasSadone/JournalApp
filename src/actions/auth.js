import { types } from "../types/types"
import { firebase, googleAuthProvider } from '../firebase/firebase-config'
import { finishLoading, startLoading } from "./ui";
import Swal from 'sweetalert2'
import { noteLogout } from "./notes";

export const startLoginEmailPassword = (email,password) => {
    return (dispatch) => {
        dispatch(startLoading())
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then( ({user}) =>{
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading())
            })
            .catch(e => {
                Swal.fire('Error', e.message, 'error')
                dispatch(finishLoading())
            })
    };
};

export const startRegisterWithEmailPasswordName = (email, password,  name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
        /*esto devuelve una promesa, que cuando tiene resultados, 
        da un user sin displayName, entonces hay que hacer esto:*/
            .then(async({user}) => {
                await user.updateProfile({displayName:name});
                dispatch(login(user.uid, user.displayName));    
            })
            //que devuelve otra promesa, por eso async await.
            .catch(e =>{
                console.log(e)
            })
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then( ({user}) => {
                dispatch(
                    login(user.uid, user.displayName)
                )
            });
    }
}

export const login = (uid,displayName) =>({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch( logout() )
        dispatch(noteLogout())
    }
}

export const logout = () => ({
    type:types.logout
})