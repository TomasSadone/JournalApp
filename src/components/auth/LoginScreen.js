import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { useDispatch, useSelector } from "react-redux";
import validator from 'validator';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';

export const LoginScreen = () => {

  const dispatch = useDispatch()
  const {loading} = useSelector(state => state.ui);

  const [formValues, handleInputChange] = useForm({
    email: 'tomas@mail.com',
    password: '123123'
  });

  const {email, password} = formValues;

  const handleLogin = (e) =>{
    e.preventDefault();
    if(isLoginValid()){
      dispatch(startLoginEmailPassword(email, password));
    }
  };

  const isLoginValid = () => {
    if (!validator.isEmail(email)) {
      return false
    } else if (password.length < 5) {
      return false
    }
    return true 
  }
  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  }

  return (
    <>
      <h3 className='auth__title'>Login</h3>

      <form onSubmit={handleLogin} className="animate__animated animate__fadeIn animate__faster">

        <input
          type='text'
          placeholder='Email'
          name='email'
          className='auth__input'
          autoComplete='off'
          value={email}
          onChange={(target)=>handleInputChange(target)}
        />

        <input
          type='password'
          placeholder='Password'
          name='password'
          className='auth__input'
          value={password}
          onChange={handleInputChange}
        />

        <button
        type='submit'
        className='btn btn-primary btn-block'
        disabled={loading}
        >
          Login
        </button>
        
        <div className='auth__social-networks'>

          <p>Login with social network</p>

          <div className="google-btn" onClick={handleGoogleLogin}>

            <div className="google-icon-wrapper">
                <img 
                  className="google-icon" 
                  src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" 
                  alt="google button" 
                />
            </div>

            <p className="btn-text">
                <b>Sign in with google</b>
            </p>

          </div>

        </div>

        <Link
          className='link' 
          to='/auth/register'>
          Create new account
        </Link>

      </form>
    </>
  )
}
