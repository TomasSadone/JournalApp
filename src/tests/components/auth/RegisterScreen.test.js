import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";

import { MemoryRouter } from 'react-router-dom';
import { mount } from "enzyme";
import { Provider } from "react-redux"
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: null
    }
}; 

let store = mockStore(initState);
// store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <RegisterScreen/>
        </MemoryRouter>
    </Provider>
);
describe('pruebas en RegisterScreen', () => { 
    test('debe hacer match con snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe hacer dispatch de accion pertinente', () => {
        const emailField = wrapper.find('input[name="email"]');
        emailField.simulate('change',{
            target: {
                value: '',
                name: 'email'
            }
        });
        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });
        const actions = store.getActions();
        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Debe ingresar un mail valido'
        });
    });
    
    test('debe mostrar caja de alerta con el error', () => {
        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Email no es correcto'
            },
            notes: {
                notes: [],
                active: null
            }
        }; 
        
        let store = mockStore(initState);
        
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen/>
                </MemoryRouter>
            </Provider>
        );
        expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initState.ui.msgError)
    })
});