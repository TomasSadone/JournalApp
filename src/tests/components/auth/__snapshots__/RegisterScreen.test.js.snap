// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en RegisterScreen debe hacer match con snapshot 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Tomas"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block"
      type="submit"
    >
      Register
    </button>
    <a
      className="link mt-5"
      href="/auth/login"
      onClick={[Function]}
    >
      Already Registred?
    </a>
  </form>,
]
`;
